c/shared/source/ufowrite/ufowrite.c

```diff
@@ -650,6 +553,219 @@ static void writeBlueValues(ufwCtx h, abfPrivateDict *privateDict){
+            fd->FontMatrix.array[5] = 0.0;
+        }
+        writeLine(h, "\t<key>FontMatrix</key>");
+        writeLine(h, "\t<array>");
+        for (i = 0; i < fd->FontMatrix.cnt; i++) {
+            float stem = fd->FontMatrix.array[i];
+            if (stem == ((int)stem))
+                sprintf(buffer, "\t\t<integer>%d</integer>", (int)stem);
+            else
+                sprintf(buffer, "\t\t<real>%.3f</real>", stem);
+            writeLine(h, buffer);
+        }
+        writeLine(h, "\t</array>");
+
+        privateDict = &(fd->Private);
+        writeLine(h, "\t<key>PrivateDict</key>");
+        writeLine(h, "\t<dict>");
+        writeBlueValues(h, privateDict);
+        writeLine(h, "\t</dict>");
+        writeLine(h, "\t</dict>");
+    }
+    writeLine(h, "\t</array>");
+}
+
+static void writeCIDMap(ufwCtx h, abfTopDict *top, char *buffer) {
+    int i;
+    writeLine(h, "\t<key>com.adobe.type.postscriptCIDMap</key>");
+    writeLine(h, "\t<dict>");
+    for (i = 0; i < h->glyphs.cnt; i++) {
+        sprintf(buffer, "\t\t<key>%s</key>", h->glyphs.array[i].glyphName);
+        writeLine(h, buffer);
+        sprintf(buffer, "\t\t<integer>%d</integer>", h->glyphs.array[i].cid);
+        writeLine(h, buffer);
+    }
+    writeLine(h, "\t</dict>");
+}
+
+static void writeGlyphOrder(ufwCtx h) {
+    char buffer[FILENAME_MAX];
+    int i;
+
+    /* Set error handler */
+    DURING_EX(h->err.env)
+
+    h->state = 1; /* Indicates writing to dst stream */
+
+    /* Open lib.plist file as dst stream */
+
+    sprintf(buffer, "%s", "lib.plist");
+    h->cb.stm.clientFileName = buffer;
+    h->stm.dst = h->cb.stm.open(&h->cb.stm, UFW_DST_STREAM_ID, 0);
+    if (h->stm.dst == NULL)
+        fatal(h, ufwErrDstStream);
+
+    writeLine(h, XML_HEADER);
+    writeLine(h, PLIST_DTD_HEADER);
+    writeLine(h, "<plist version=\"1.0\">");
+    writeLine(h, "<dict>");
+    writeLine(h, "\t<key>public.glyphOrder</key>");
+    writeLine(h, "\t<array>");
+    for (i = 0; i < h->glyphs.cnt; i++) {
+        Glyph *glyphRec;
+        glyphRec = &h->glyphs.array[i];
+        sprintf(buffer, "\t\t<string>%s</string>", glyphRec->glyphName);
+        writeLine(h, buffer);
+    }
+
+    writeLine(h, "\t</array>");
+
+    if (h->top->sup.flags & ABF_CID_FONT) {
+        if (h->top->cid.CIDFontName.ptr != NULL) {
+            writeLine(h, "\t<key>com.adobe.type.CIDFontName</key>");
+            sprintf(buffer, "\t<string>%s</string>", h->top->cid.CIDFontName.ptr);
+            writeLine(h, buffer);
+            writeLine(h, "\t<key>com.adobe.type.ROS</key>");
+            sprintf(buffer, "\t<string>%s-%s-%ld</string>", h->top->cid.Registry.ptr, h->top->cid.Ordering.ptr, h->top->cid.Supplement);
+            writeLine(h, buffer);
+        }
+        writeFDArray(h, h->top, buffer);
+        writeCIDMap(h, h->top, buffer);
+    }
+    writeLine(h, "</dict>");
+    writeLine(h, "</plist>");
+
+    /* Close dst stream */
+    flushBuf(h);
```

Comment:
I'm worried we may have missed a significant aspect of this implementation. 

One of the reasons we chose to move these structures from `fontinfo.plist` to `lib.plist` is that the keys in the former are fixed and implementations throw other keys away. If we want our custom data to stay we need to use `lib.plist` (and `groups.plist`) instead. 

However, if someone is using `tx` to take a round-trip from UFO to UFO we need to honor the same convention: if there's a key in `lib.plist` that isn't one of ours it needs to be preserved. That probably means opening the file with libxml2, modifying that DOM, and then outputting it via libxml2 as well. If the file isn't there we can have this writeLine based stuff as an alternative or (probably preferable) start with a blank LibXML2 DOM object and add our elements to it.

Trying to go through our custom stream IO crap (I really hate it) to do that is probably not worth it, especially given that we're opening new files here anyway, so I would just do the most natural LibXML2-flavored thing.

Reply:
Yes, we want to preserve existing data in a UFO when possible. If we go UFO > t1 then we obviously can't round trip all the data because there is nowhere to store it. But if we want to, for example, subset a CID-keyed UFO then anything in lib.plist not ours should be retained while our keys should be updated. Doing a basic `tx -ufo -o newfile.ufo oldfile.ufo` should preserve everything because nothing else has happened. Something like `tx -ufo -o newfilesubset.ufo oldfile.ufo -g 0-44` should also remove everything except these glyphs from the UFO, public.glyphOrder, CIDMap, and groups.plist.

Reply:
An implementation for all UFO keys to be preserved (including unidentified keys) for UFO -> UFO roundtripping will come in a future PR made after this one merges.